commons/all/src/main/java/org/infinispan/commons/io/FileWatcher.java

```diff
@@ -0,0 +1,116 @@
+package org.infinispan.commons.io;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+import org.infinispan.commons.logging.Log;
+import org.infinispan.commons.logging.LogFactory;
+
+/**
+ * @since 15.0
+ */
+public class FileWatcher implements Runnable {
+   private static final Log log = LogFactory.getLog(FileWatcher.class);
+   public static final int SLEEP = 2000;
+   private static FileWatcher INSTANCE;
+   private final Thread thread;
+   private final ConcurrentHashMap<Path, Watched> watched;
+   private volatile boolean running = true;
+
+   public static synchronized void watch(Path path, Consumer<Path> callback) throws IOException {
+      if (INSTANCE == null) {
```

Comment:
If we keep this class, this can be done with initialization on demand idiom or enum singleton instead, removing the need for synchronization (especially since you are using a CHM).